File: sdk-java/hfcx-sdk-core/src/main/java/eg/gov/healthflow/hfcx/sdk/core/validators/EgyptianPhoneValidator.java
package eg.gov.healthflow.hfcx.sdk.core.validators;

import java.util.regex.Pattern;

/**
 * Validator and normaliser for Egyptian mobile phone numbers.
 *
 * <p>Accepts the four common forms:
 *
 * <ul>
 *   <li>{@code +201[0125]XXXXXXXX} — international with leading {@code +}</li>
 *   <li>{@code 00201[0125]XXXXXXXX} — international with double-zero prefix</li>
 *   <li>{@code 201[0125]XXXXXXXX} — bare country code</li>
 *   <li>{@code 01[0125]XXXXXXXX} — local form with leading zero</li>
 * </ul>
 *
 * <p>Mobile network prefixes are {@code 010} (Vodafone), {@code 011}
 * (Etisalat), {@code 012} (Orange), {@code 015} (WE).
 */
public final class EgyptianPhoneValidator {

    private static final Pattern NORMALISED = Pattern.compile("^\\+201[0125]\\d{8}$");

    private EgyptianPhoneValidator() {}

    public static boolean isValid(String phone) {
        return normalise(phone) != null;
    }

    /**
     * @return the canonical {@code +201XXXXXXXXX} form, or {@code null} if
     *     {@code phone} is not a valid Egyptian mobile number.
     */
    public static String normalise(String phone) {
        if (phone == null) {
            return null;
        }
        String trimmed = phone.trim().replace(" ", "").replace("-", "");
        String candidate;
        if (trimmed.startsWith("+20")) {
            candidate = trimmed;
        } else if (trimmed.startsWith("0020")) {
            candidate = "+20" + trimmed.substring(4);
        } else if (trimmed.startsWith("20") && trimmed.length() == 12) {
            candidate = "+" + trimmed;
        } else if (trimmed.startsWith("0") && trimmed.length() == 11) {
            candidate = "+20" + trimmed.substring(1);
        } else {
            return null;
        }
        return NORMALISED.matcher(candidate).matches() ? candidate : null;
    }
}

File: sdk-dotnet/src/HealthFlow.Hfcx.Sdk/Validators/EgyptianPhoneValidator.cs
// Copyright (c) HealthFlow Medical HCX. Licensed under Apache 2.0.

using System;
using System.Text.RegularExpressions;

namespace HealthFlow.Hfcx.Sdk.Validators;

/// <summary>
/// Validator and normaliser for Egyptian mobile phone numbers. Cross-SDK
/// invariant with Java's <c>EgyptianPhoneValidator</c> and Python's
/// <c>egyptian_phone</c>: accepts the four canonical mobile forms
/// (international <c>+</c>, double-zero, bare country code, local
/// 0-prefixed) and returns the <c>+201XXXXXXXXX</c> canonical form.
/// Mobile prefixes are 010, 011, 012, 015 (Vodafone, Etisalat, Orange,
/// WE).
/// </summary>
public static partial class EgyptianPhoneValidator
{
    /// <summary>Returns <see langword="true"/> iff the phone is a valid Egyptian mobile.</summary>
    public static bool IsValid(string? phone) => Normalise(phone) is not null;

    /// <summary>Return the canonical <c>+201XXXXXXXXX</c> form, or <see langword="null"/>.</summary>
    public static string? Normalise(string? phone)
    {
        if (phone is null)
        {
            return null;
        }

        var trimmed = phone.Trim().Replace(" ", "", StringComparison.Ordinal).Replace("-", "", StringComparison.Ordinal);
        string candidate;
        if (trimmed.StartsWith("+20", StringComparison.Ordinal))
        {
            candidate = trimmed;
        }
        else if (trimmed.StartsWith("0020", StringComparison.Ordinal))
        {
            candidate = "+20" + trimmed[4..];
        }
        else if (trimmed.StartsWith("20", StringComparison.Ordinal) && trimmed.Length == 12)
        {
            candidate = "+" + trimmed;
        }
        else if (trimmed.StartsWith('0') && trimmed.Length == 11)
        {
            candidate = "+20" + trimmed[1..];
        }
        else
        {
            return null;
        }

        return Normalised().IsMatch(candidate) ? candidate : null;
    }

    [GeneratedRegex(@"^\+201[0125]\d{8}$")]
    private static partial Regex Normalised();
}
